godot-core/src/engine/file.rs

```diff
@@ -0,0 +1,766 @@
+                let new_pos = new_pos as u64;
+                self.fa.seek(new_pos);
+                Ok(new_pos)
+            }
+        }
+    }
+}
+
+impl BufRead for GFile {
+    fn fill_buf(&mut self) -> IoResult<&[u8]> {
+        // We need to determine number of remaining bytes - otherwise the `FileAccess::get_buffer return in an error`
+        let remaining_bytes = self.fa.get_length() - self.fa.get_position();
+        let buffer_read_size = cmp::min(remaining_bytes as usize, Self::BUFFER_SIZE);
+        // We need to keep the amount of last read side to be able to adjust cursor position in `consume`
+        self.last_buffer_size = buffer_read_size;
+
+        self.buffer = vec![0; Self::BUFFER_SIZE];
+
+        let gd_buffer = self.fa.get_buffer(buffer_read_size as i64);
+        self.check_error()?;
+
+        for i in 0..gd_buffer.len() {
+            self.buffer[i] = gd_buffer.get(i);
+        }
+
+        Ok(&self.buffer[0..gd_buffer.len()])
+    }
+
+    fn consume(&mut self, amt: usize) {
+        // Cursor is being moved by `FileAccess::get_buffer()` call, so we need to adjust it
+        self.seek(std::io::SeekFrom::Current(
+            -((self.last_buffer_size - amt) as i64),
+        ))
+        .unwrap();
+    }
+}
+
+/// Error stemming from the [Gd] not being unique.
+///
+/// Keeping track of uniqueness of the references can crucial in many applications, especially if we
+/// want to make sure that the passed [Gd] object will be in possession of only one different
+/// object instance or function in their lifetime.
+///
+/// Only applicable to [GodotClass](godot::obj::GodotClass) objects that inherits from [RefCounted].
+/// To check the uniqueness, call the `check()` associated method.
+///
+/// ## Example
+///
+/// ```no_run
+/// let node_reference = Node::new_alloc();
+/// let cloned_reference = node_reference.clone();
+/// let result = GdRefUniqueness::check(node_reference);
+///
+/// assert!(result.is_err());
+///
+/// if let Err(error) = result {
+///     assert_eq!(error.get_reference_count(), 2)
+/// }
+/// ```
+pub struct GdRefUniqueness {
+    reference_count: i32,
+}
+
+impl GdRefUniqueness {
+    /// check [Gd] reference uniqueness.
+    ///
+    /// Checks the [Gd] of the [GodotClass](godot::obj::GodotClass) that inherits from [RefCounted] if it is unique
+    /// reference to the object.
+    ///
+    /// ## Example
+    ///
+    /// ```no_run
+    /// use godot::engine::Node;
+    ///
+    /// let unique_reference = Node::new_alloc();
+    /// let unique_result = GdRefUniqueness::check(unique_reference);
+    ///
+    /// let shared_reference = Node::new_alloc();
+    /// let cloned_reference = shared_reference.clone();
+    /// let shared_result = GdRefUniqueness::check(shared_reference);
+    /// let cloned_result = GdRefUniqueness::check(cloned_reference);
+    ///
+    /// assert!(unique_result.is_ok());
+    /// assert!(shared_result.is_err());
+    /// assert!(cloned_result.is_err());
```

Comment:
This does not work because `Node` is not reference-counted. You would need to limit the bounds to something that's `RefCounted`, but it's a bit complex to state; a TODO is fine.

For now, `RefCounted` instead of `Node` in the example should be good.

I would do checks immediately after the variables, to avoid scopes spanning too far. Also names can be shorter.
```suggestion
    /// let unique = RefCounted::new_alloc();
    /// assert!(GdRefUniqueness::check(unique).is_ok());
    ///
    /// let origin = RefCounted::new();
    /// let cloned = origin.clone();
    /// assert!(GdRefUniqueness::check(origin).is_err());
    /// assert!(GdRefUniqueness::check(cloned).is_err());
```